cmd/controller-manager/main.go

```diff
@@ -144,6 +213,144 @@ func main() {
 	setupLog.Info("Got config for ITS", "name", itsName)
 	itsRestConfig = itsClientLimits.LimitConfig(itsRestConfig)
 
+	// Start /readyz endpoint - be resilient to ITS connection issues
+	go func() {
+		http.HandleFunc("/readyz", func(w http.ResponseWriter, r *http.Request) {
+			// Controller-manager is ready if:
+			// 1. It has become leader (if leader election is enabled), OR
+			// 2. It's running without leader election (if leader election is disabled)
+			// 3. AND it has successfully connected to WDS at least once
+			if readyFlag.Load() {
+				w.WriteHeader(http.StatusOK)
+				w.Write([]byte("ok"))
+			} else {
+				w.WriteHeader(http.StatusServiceUnavailable)
+				w.Write([]byte("not ready - waiting for leadership or controller startup"))
+			}
+		})
+		http.ListenAndServe(":8081", nil)
+	}()
+
+	// Start controllers based on leader election setting
+	if enableLeaderElection {
+		setupLog.Info("Starting with leader election enabled")
+		startControllersWithLeaderElection(ctx, setupLog, wdsRestConfig, itsRestConfig, wdsName, itsName, allowedGroupsSet, ctlrsToStart, wdsClientMetrics, itsClientMetrics)
+	} else {
+		setupLog.Info("Starting without leader election")
+		startControllersDirectly(ctx, setupLog, wdsRestConfig, itsRestConfig, wdsName, itsName, allowedGroupsSet, ctlrsToStart, wdsClientMetrics, itsClientMetrics)
+	}
+
+	select {}
+}
+
+// startControllersWithLeaderElection uses client-go leader election instead of controller-runtime
+func startControllersWithLeaderElection(ctx context.Context, setupLog klog.Logger, wdsRestConfig, itsRestConfig *rest.Config, wdsName, itsName string, allowedGroupsSet sets.Set[string], ctlrsToStart sets.Set[string], wdsClientMetrics, itsClientMetrics ksmetrics.ClientMetrics) {
+	// Create WDS client for leader election with retry for transient errors
+	var wdsClient *kubernetes.Clientset
+	var err error
+	for i := 0; i < 3; i++ {
+		wdsClient, err = kubernetes.NewForConfig(wdsRestConfig)
+		if err == nil {
+			break
+		}
+		if i < 2 && isTransientError(err) {
+			setupLog.Info("Failed to create WDS client due to transient error, retrying", "attempt", i+1, "error", err)
+			time.Sleep(time.Duration(i+1) * time.Second)
+		} else if i < 2 {
+			setupLog.Info("Failed to create WDS client, retrying", "attempt", i+1, "error", err)
+			time.Sleep(time.Duration(i+1) * time.Second)
+		}
+	}
+	if err != nil {
+		setupLog.Error(err, "unable to create WDS client for leader election after retries")
+		os.Exit(1)
+	}
+
+	// Check/create namespace for leader election with retry for transient errors
+	var namespaceErr error
+	for i := 0; i < 3; i++ {
```

Comment:
Why retry a limited number of times?
Why not use one of the Kubernetes utility functions that retries with exponential backoff (see https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait)?